### PR TITLE
BLL layer for user entity

### PR DESCRIPTION
**Pull request by KiSSel3**
- Created user dto
- Added UserService interface and implementation
- Validators have been created for user dto
- Configured mapper for user dto

### CODE REVIEW

File: OnlineShop/OnlineShop.BLL/Mappers/MapsterConfig.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    public static void Configure()
+    {
+        TypeAdapterConfig<string, Guid>.NewConfig()
+            .MapWith(src => StringToGuidMapper.ToGuid(src));
+
+        TypeAdapterConfig<Guid,string>.NewConfig()
+            .MapWith(src => StringToGuidMapper.ToString(src));
+        
+        TypeAdapterConfig<RegisterRequestDTO, User>.NewConfig()
+            .Map(dest => dest.PasswordHash, src => PasswordHelper.HashPassword(src.Password))
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Mappers/MapsterConfig.cs]: This call must take place inside the service:PasswordHelper.HashPassword(src.Password)

File: OnlineShop/OnlineShop.BLL/Mappers/MapsterConfig.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+public static class MapsterConfig
+{
+    public static void Configure()
+    {
+        TypeAdapterConfig<string, Guid>.NewConfig()
+            .MapWith(src => StringToGuidMapper.ToGuid(src));
+
+        TypeAdapterConfig<Guid,string>.NewConfig()
+            .MapWith(src => StringToGuidMapper.ToString(src));
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Mappers/MapsterConfig.cs]: I don't understand why you need this. just call inside the services and that's it

File: OnlineShop/OnlineShop.BLL/Mappers/MapsterConfig.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        
+        TypeAdapterConfig<RegisterRequestDTO, User>.NewConfig()
+            .Map(dest => dest.PasswordHash, src => PasswordHelper.HashPassword(src.Password))
+            .Map(dest => dest.IsDeleted, src => false);
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Mappers/MapsterConfig.cs]: why should this field be set like this?

File: OnlineShop/OnlineShop.BLL/Mappers/StringToGuidMapper.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    public static Guid ToGuid(string value)
+    {
+        if (string.IsNullOrEmpty(value))
+        {
+            throw new ArgumentNullException(nameof(value));
+        }
+
+        if (!Guid.TryParse(value, out var guid))
+        {
+            throw new MappingException($"The string '{value}' is not a valid Guid.");
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Mappers/StringToGuidMapper.cs]: This is not an exception to mapping, but to converting values

File: OnlineShop/OnlineShop.BLL/Mappers/StringToGuidMapper.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        }
+
+        return guid;
+    }
+
+    public static string ToString(Guid value)
+    {
+        return value.ToString();
+    }
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Mappers/StringToGuidMapper.cs]: Why do you need this

File: OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs [hunk 6]
```diff
@@ -1,8 +1,73 @@
...
 
-public class UserService:IUserService
+public class UserService : IUserService
 {
-    
+    private readonly IUserRepository _userRepository;
+
+    public UserService(IUserRepository userRepository)
+    {
+        MapsterConfig.Configure();
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs]: The configuration can be moved outside of this class and registered where the di registration takes place

File: OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs [hunk 7]
```diff
@@ -1,8 +1,73 @@
...
+        
+        _userRepository = userRepository;
+    }
+
+    public async Task<UserResponseDTO> GetByIdAsync(string userId, CancellationToken cancellationToken = default)
+    {
+        var id = userId.Adapt<Guid>();
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs]: redo

File: OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs [hunk 8]
```diff
@@ -1,8 +1,73 @@
...
+        }
+
+        var userResponseDTO = user.Adapt<UserResponseDTO>();
+
+        return userResponseDTO;
+    }
+
+    public async Task LoginAsync(LoginRequestDTO loginRequestDTO, CancellationToken cancellationToken = default)
+    {
+        LoginValidator loginValidator = new LoginValidator();
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs]: use var here

File: OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs [hunk 9]
```diff
@@ -1,8 +1,73 @@
...
+        await loginValidator.ValidateAndThrowAsync(loginRequestDTO, cancellationToken);
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs]: And why do you need to create a validator in general, you can use auto-validation

File: OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs [hunk 10]
```diff
@@ -1,8 +1,73 @@
...
+        if(!PasswordHelper.VerifyPassword(user.PasswordHash, loginRequestDTO.Password))
+        {
+            throw new AuthenticationException("Login or password entered incorrectly.");
+        }
+    }
+
+    public async Task RegisterAsync(RegisterRequestDTO registerRequestDTO, CancellationToken cancellationToken = default)
+    {
+        RegisterValidator registerValidator = new RegisterValidator();
+        await registerValidator.ValidateAndThrowAsync(registerRequestDTO, cancellationToken);
```
Comment (mebry) [inline OnlineShop/OnlineShop.BLL/Services/Implementations/UserService.cs]: the same

**Review by mebry — CHANGES_REQUESTED** [review]

**Comment by KiSSel3** [comment]
Corrected errors

**Review by mebry — APPROVED** [review]

**Pull request merged**